lib/parsers/wav_parser.rb

```diff
@@ -20,43 +20,39 @@ def call(io)
     # with the exception that the Format chunk must precede the Data chunk.
     # The specification does not require the Format chunk to be the first chunk
     # after the RIFF header.
-    # http://soundfile.sapp.org/doc/WaveFormat/
-    # For WAVE files containing PCM audio format we parse the 'fmt ' and
-    # 'data' chunks while for non PCM audio formats the 'fmt ' and 'fact'
-    # chunks. In the latter case the order fo appearence of the chunks is
-    # arbitrary.
-    fmt_processed = false
-    fact_processed = false
+    # https://www.mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/WAVE.html
     fmt_data = {}
-    total_sample_frames = 0
+    data_size = 0
+    total_sample_frames = nil
     loop do
       chunk_type, chunk_size = safe_read(io, 8).unpack('a4l')
       case chunk_type
       when 'fmt ' # watch out: the chunk ID of the format chunk ends with a space
         fmt_data = unpack_fmt_chunk(io, chunk_size)
-        return process_non_pcm(fmt_data, total_sample_frames) if fmt_data[:audio_format] != 1 and fact_processed
-        fmt_processed = true
       when 'data'
-        return unless fmt_processed # the 'data' chunk cannot preceed the 'fmt ' chunk
-        return process_pcm(fmt_data, chunk_size) if fmt_data[:audio_format] == 1
-        safe_skip(io, chunk_size)
+        data_size = chunk_size
       when 'fact'
         total_sample_frames = safe_read(io, 4).unpack('l').first
         safe_skip(io, chunk_size - 4)
-        return process_non_pcm(fmt_data, total_sample_frames) if fmt_processed and fmt_data[:audio_format] != 1
-        fact_processed = true
       else
         # Skip this chunk until a known chunk is encountered
         safe_skip(io, chunk_size)
       end
+    rescue FormatParser::IOUtils::InvalidRead
+      # We've reached EOF, so it's time to make the most out of the metadata we
+      # managed to parse
+      break
     end
+
+    file_info(fmt_data, data_size, total_sample_frames)
```

Comment:
Previously we returned from inside the loop depending on some conditions. Is there any issue with potentially having multiple chunks of the same type and then only returning for the final one of these?

Reply:
As far as I got it, chunks can be unordered (more or less), but they cannot repeat themselves (at least, not the 3 ones referred to in this PR). That would lead to a pretty invalid file. And if that's the case, I don't think it'll matter much what we return.

Reply:
Should we at least check that `fmt_data` exists as otherwise we access it inside file_info when it's potentially nil (which was previously impossible) and avoid random nil access errors?